### PR TITLE
VGG Draft Outline 

### PR DESCRIPTION
**Pull request by albertkao227**
This is my outline for introducing VGG network. One sentence summary was given for each point for further expansion. I omitted completing the code for we may want to do that in the next step. Please let me know if there is any feedback.

### CODE REVIEW

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+|Conv2D|3X3|512X28X28|2,359,808|
+|Conv2D|3X3|512X28X28|2,359,808|
+|MaxPool2D|2X2|512X14X14|0|
+|Conv2D|3X3|512X14X14|2,359,808|
+|Conv2D|3X3|512X14X14|2,359,808|
+|Conv2D|3X3|512X14X14|2,359,808|
+|MaxPool2D|2X2|0|
+|Linear|N/A|4096|102,764,544|
+|Linear|N/A|4096|16,781,312|
+|Linear|N/A|1000|4,096,000|
```
Comment (sezan92) [inline course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md:44]: I find making these like block diagram would be better. no need to change in this PR. we can make another PR after merging this one, for adding a diagram. what do you think @albertkao227 ?

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+
+
+3. Key Comparisons 
+
+- VGG (16 or 19 layers) is relatively deeper than other SOTA networks at the time. AlexNet, winning model for ILSVRC 2012 only has 8 layers.    
+- Multiple small (3X3) receptive field filters with ReLU activation instead of one large (7X7 or 11X11) filter lead to better learning of complex features. Smaller filters also means fewer parameters per layer, with additional nonlinearity introduced in between.
```
Comment (sezan92) [inline course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md:51]: should not we use an explanatory diagram for this? i am just asking question.

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
+
+# Very Deep Convolutional Networks for Large Scale Image Recognition (2014)
+
+
+1. Introduction
+
+VGG architecture was developed in 2014 by Karen Simonyan and Andrew Zisserman from Visual Geometry Group at Oxford University. The authors detailed key advancements in convolutional neural networks for ImageNet Large Scale Visual Recognition Challenge (ILSVRC). VGG demonstrated significant improvements with increased depth and repeated small (3X3) convolutional filters. It was a research project to systematically evaluate prior-art configurations and identify key structural designs.
```
Comment (sezan92) [inline course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md:7]:
```suggestion
The VGG architecture was developed in 2014 by Karen Simonyan and Andrew Zisserman from the Visual Geometry Group at Oxford University. The authors detailed key advancements in convolutional neural networks for the ImageNet Large Scale Visual Recognition Challenge (ILSVRC). VGG demonstrated significant improvements with increased depth and repeated small (3X3) convolutional filters. It was a research project to systematically evaluate prior art configurations and identify key structural designs. 
```

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+2. VGG Network Architechture 
+   
+- Input is 224X224 image. 
+- 3X3 convolution kernels with 2X2 max pooling layers.
+- Number of channels for each convolutional layer 64 -> 128 -> 256 -> 512 -> 512. 
+- VGG16 has 16 hidden layers (13 convolutional layers and 3 fully connected layers).
+- VGG19 has 19 hidden layers (16 convolutional layers and 3 fully connected layers).
+
+Below is network structure of VGG16, The activation and dropout layers were omitted for brevity. MaxPooling layers were included to display change in output tensor size at each layer.
```
Comment (sezan92) [inline course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md:19]:
```suggestion
2. VGG Network Architechture 
   
- Input is 224X224 image. 
- 3X3 convolution kernels with 2X2 max pooling layers.
- Number of channels for each convolutional layer 64 -> 128 -> 256 -> 512 -> 512. 
- VGG16 has 16 hidden layers (13 convolutional layers and 3 fully connected layers).
- VGG19 has 19 hidden layers (16 convolutional layers and 3 fully connected layers).

Below is the network structure of VGG16, The activation and dropout layers were omitted for brevity. MaxPooling layers were included to display changes in output tensor size at each layer. 
```

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md (hunk 2 above)
Comment (sezan92) [reply course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md:51]:
```suggestion
- Multiple small (3X3) receptive field filters with ReLU activation instead of one large (7X7 or 11X11) filter lead to better learning of complex features. Smaller filters also mean fewer parameters per layer, with additional nonlinearity introduced in between.   
```

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+|Linear|N/A|1000|4,096,000|
+
+
+
+3. Key Comparisons 
+
+- VGG (16 or 19 layers) is relatively deeper than other SOTA networks at the time. AlexNet, winning model for ILSVRC 2012 only has 8 layers.    
+- Multiple small (3X3) receptive field filters with ReLU activation instead of one large (7X7 or 11X11) filter lead to better learning of complex features. Smaller filters also means fewer parameters per layer, with additional nonlinearity introduced in between.   
+- Multiscale training and inference. Each image was trained multiple rounds with varying scales to ensure similar characteristics are captured at different sizes.  
+- Consistency and simplicity of VGG network makes it easier to scale or modify for future improvements.
```
Comment (sezan92) [inline course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md:53]:
```suggestion
- Multiscale training and inference. Each image was trained in multiple rounds with varying scales to ensure similar characteristics were captured at different sizes.  
- Consistency and simplicity of the VGG network make it easier to scale or modify for future improvements. 
```

**Review by sezan92 — COMMENTED** [review]
commented

**Review by ShamieCC — APPROVED** [review]
We would have to add the explanatory diagram as an addition to what is being explained. I am happy with everything so far and the suggestions from @sezan92. We do have to add basic information about the model  such as advantages and disadvantages of vgg19 and common use cases. Not sure if there are others covering a topic that covers this basic information, so that we focus on the technical aspects. But if not, we could add the following:

The VGG19 network has several advantages over other CNN architectures, including:

Accuracy: The network achieved state-of-the-art results on the ILSVRC 2014 challenge, with a top-5 error rate of 22.7%.
Simplicity: The architecture is relatively simple to implement and train.
Efficiency: The network can be efficiently run on GPUs, making it suitable for real-time applications.

However, the VGG19 network also has some disadvantages:

Size: The network is a very large and complex model, with over 140 million parameters. This can make it difficult to train and deploy on some devices.
Computational cost: The network is computationally expensive to run, especially on CPUs.
Can be prone to overfitting on small datasets

Overall, the network is a very powerful and accurate CNN architecture. It has been used to achieve state-of-the-art results on a wide range of image recognition tasks.

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md (hunk 1 above)
Comment (albertkao227) [reply course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md:44]: @sezan92  Yes I believe a diagram is better. We should make it or find a good one and reference.

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md (hunk 2 above)
Comment (albertkao227) [reply course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md:51]: @sezan92  are you thinking about showing AlexNet vs VGG side by side, and maybe point out the filter size differences? Yes that is a great idea.

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md (hunk 4 above)
Comment (albertkao227) [reply course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md:19]: looks good

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md (hunk 3 above)
Comment (albertkao227) [reply course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md:7]: looks good

**Review by JvThunder — APPROVED** [review]
As Shamie stated, I feel like including an architecture diagram and the advantages and disadvantages of VGG compared to previous works will certainly help make it more accessible and give better insights for beginners. In addition, I feel like we could also include how simply adding more layers would not suffice for significant improvement, which would require research of novel architectures beyond simply increasing the number of layers. (We can include this here or when discussing the next architecture.)

Also, if we end up using the block diagram, I noticed that the final MaxPool layer is written as "|MaxPool2D|2X2|0|". I feel like you missed the layer output size, and it should've been "|MaxPool2D|2X2|512X7X7|0|" instead.

Overall, it's a great summary so I will approve the changes for now.

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+3. Key Comparisons 
+
+- VGG (16 or 19 layers) is relatively deeper than other SOTA networks at the time. AlexNet, winning model for ILSVRC 2012 only has 8 layers.    
+- Multiple small (3X3) receptive field filters with ReLU activation instead of one large (7X7 or 11X11) filter lead to better learning of complex features. Smaller filters also means fewer parameters per layer, with additional nonlinearity introduced in between.   
+- Multiscale training and inference. Each image was trained multiple rounds with varying scales to ensure similar characteristics are captured at different sizes.  
+- Consistency and simplicity of VGG network makes it easier to scale or modify for future improvements. 
+
+
+4. Pytorch Example
```
Comment (salgadev) [inline course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md:56]: Missing the example and I think there should be some connection to HF as in how to use with the transformers library

Comment (albertkao227) [reply course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md:56]: Yes, we should add an example and how to use it with HF in the next iteration.

Comment (sezan92) [reply course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md:56]: @socd06 i think we will add example in the next PR.

File: course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md (hunk 2 above)
Comment (sezan92) [reply course/2 Neural Network Architectures/2.1 Convolutional Neural Networks/2.1.2 Common Pre-Trained Models/vgg.md:51]: @albertkao227 although i am not entirely sure how to show the advantages in graphic way. we can consider it later. please commit the suggestion

**Review by sezan92 — APPROVED** [review]
